File: dart/packages/verdict_rules/lib/src/engine.dart
import 'result.dart';
import 'rule.dart';

/// Holds a set of rules and answers questions about them.
///
/// The engine's run modes never short-circuit.
class RulesEngine<TContext> {
  final List<Rule<TContext>> _rules;
  final Map<String, Rule<TContext>> _byName;
  final Map<String, List<Rule<TContext>>> _byGroup;

  RulesEngine(List<Rule<TContext>> rules)
      : _rules = List.unmodifiable(rules),
        _byName = {for (final r in rules) r.name: r},
        _byGroup = _indexByGroup(rules);

  static Map<String, List<Rule<TContext>>> _indexByGroup<TContext>(
      List<Rule<TContext>> rules) {
    final byGroup = <String, List<Rule<TContext>>>{};
    for (final rule in rules) {
      final group = rule.group;
      if (group != null && group.isNotEmpty) {
        byGroup.putIfAbsent(group, () => <Rule<TContext>>[]).add(rule);
      }
    }
    return byGroup;
  }

  /// Every rule name registered here, in registration order.
  Iterable<String> get ruleNames => _byName.keys;

  /// Every group label carried by at least one rule, in first-seen order.
  Iterable<String> get groupNames => _byGroup.keys;

  /// Evaluate every registered rule. Never short-circuits.
  Future<RunResult> runAll(TContext context) async {
    final results = <RuleResult>[];
    for (final rule in _rules) {
      results.add(await rule.evaluate(context));
    }
    return RunResult(passed: results.every((r) => r.passed), results: results);
  }

  /// Evaluate one rule by name, or return null if no such rule exists.
  ///
  /// Null means absent, never failed.
  Future<RuleResult?> tryRunNamed(
    String name,
    TContext context,
  ) async {
    final rule = _byName[name];
    if (rule == null) return null;
    return rule.evaluate(context);
  }

  /// Evaluate exactly one rule, looked up by name.
  ///
  /// Throws [ArgumentError] if no rule has this name.
  Future<RuleResult> runNamed(String name, TContext context) async {
    final result = await tryRunNamed(name, context);
    if (result == null) {
      throw ArgumentError.value(name, 'name', 'No rule with this name');
    }
    return result;
  }

  /// Evaluate a group, or return null if no such group exists.
  ///
  /// Null means absent, never vacuously passed.
  Future<RunResult?> tryRunGroup(
    String group,
    TContext context,
  ) async {
    final rules = _byGroup[group];
    if (rules == null || rules.isEmpty) return null;
    final results = <RuleResult>[];
    for (final rule in rules) {
      results.add(await rule.evaluate(context));
    }
    return RunResult(passed: results.every((r) => r.passed), results: results);
  }

  /// Evaluate every rule sharing a group label. Never short-circuits.
  ///
  /// Throws [ArgumentError] if no rule carries this label.
  Future<RunResult> runGroup(String group, TContext context) async {
    final result = await tryRunGroup(group, context);
    if (result == null) {
      throw ArgumentError.value(group, 'group', 'No rules in this group');
    }
    return result;
  }
}

File: js/packages/verdict-rules/src/engine.ts
import { UnknownLookupError } from "./errors.js";
import type { RuleResult, RunResult } from "./result.js";
import type { Rule } from "./rule.js";

/**
 * Holds a set of rules and answers questions about them.
 *
 * The engine's run modes never short-circuit; a composite reaches one
 * verdict and stops.
 *
 * Generic over `TContext`, the same way {@link Rule} is, with no default
 * type parameter — dict-context is `RulesEngine<Context>`.
 */
export class RulesEngine<TContext> {
  readonly #rules: readonly Rule<TContext>[];
  readonly #byName: Map<string, Rule<TContext>>;
  readonly #byGroup: Map<string, Rule<TContext>[]>;

  constructor(rules: readonly Rule<TContext>[]) {
    this.#rules = [...rules];
    this.#byName = new Map(rules.map((r) => [r.name, r]));
    this.#byGroup = new Map();
    for (const rule of rules) {
      if (rule.group === undefined || rule.group === "") continue;
      const existing = this.#byGroup.get(rule.group);
      if (existing === undefined) this.#byGroup.set(rule.group, [rule]);
      else existing.push(rule);
    }
  }

  /** Every rule name registered here, in registration order. */
  get ruleNames(): readonly string[] {
    return [...this.#byName.keys()];
  }

  /** Every group label carried by at least one rule, in first-seen order. */
  get groupNames(): readonly string[] {
    return [...this.#byGroup.keys()];
  }

  /** Evaluate every registered rule. Never short-circuits. */
  async runAll(context: TContext): Promise<RunResult> {
    const results: RuleResult[] = [];
    for (const rule of this.#rules) {
      results.push(await rule.evaluate(context));
    }
    return { passed: results.every((r) => r.passed), results };
  }

  /**
   * Evaluate one rule by name, or return `undefined` if no such rule exists.
   *
   * `undefined` means absent, never failed.
   */
  async tryRunNamed(
    name: string,
    context: TContext,
  ): Promise<RuleResult | undefined> {
    const rule = this.#byName.get(name);
    if (rule === undefined) return undefined;
    return rule.evaluate(context);
  }

  /**
   * Evaluate exactly one rule, looked up by name.
   *
   * @throws {UnknownLookupError} if no rule has this name.
   */
  async runNamed(name: string, context: TContext): Promise<RuleResult> {
    const result = await this.tryRunNamed(name, context);
    if (result === undefined) {
      throw new UnknownLookupError("rule", name);
    }
    return result;
  }

  /**
   * Evaluate a group, or return `undefined` if no such group exists.
   *
   * `undefined` means absent, never vacuously passed.
   */
  async tryRunGroup(
    group: string,
    context: TContext,
  ): Promise<RunResult | undefined> {
    const rules = this.#byGroup.get(group);
    if (rules === undefined || rules.length === 0) return undefined;
    const results: RuleResult[] = [];
    for (const rule of rules) {
      results.push(await rule.evaluate(context));
    }
    return { passed: results.every((r) => r.passed), results };
  }

  /**
   * Evaluate every rule sharing a group label. Never short-circuits.
   *
   * @throws {UnknownLookupError} if no rule carries this label.
   */
  async runGroup(group: string, context: TContext): Promise<RunResult> {
    const result = await this.tryRunGroup(group, context);
    if (result === undefined) {
      throw new UnknownLookupError("group", group);
    }
    return result;
  }
}
